File: src/main/scala/blog/typed/scaladsl/Greeter1.scala
package blog.typed.scaladsl

import scala.io.StdIn

import akka.typed.ActorSystem
import akka.typed.Behavior
import akka.typed.scaladsl.Actor

object Greeter1 {
  sealed trait Command
  case object Greet extends Command
  final case class WhoToGreet(who: String) extends Command

  val greeterBehavior: Behavior[Command] =
    Actor.mutable[Command](ctx => new Greeter1)
}

class Greeter1 extends Actor.MutableBehavior[Greeter1.Command] {
  import Greeter1._

  private var greeting = "hello"

  override def onMessage(msg: Command): Behavior[Command] = {
    msg match {
      case WhoToGreet(who) =>
        greeting = s"hello, $who"
      case Greet =>
        println(greeting)
    }
    this
  }
}

object HelloWorldApp1 {
  def main(args: Array[String]): Unit = {
    val root = Actor.deferred[Nothing] { ctx =>
      import Greeter1._
      val greeter = ctx.spawn(greeterBehavior, "greeter")
      greeter ! WhoToGreet("World")
      greeter ! Greet

      Actor.empty
    }
    val system = ActorSystem[Nothing]("HelloWorld", root)
    try {
      println("Press ENTER to exit the system")
      StdIn.readLine()
    } finally {
      system.terminate()
    }
  }
}

File: src/main/java/blog/typed/javadsl/Greeter1.java
package blog.typed.javadsl;

import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.Actor;
import akka.actor.typed.javadsl.Actor.MutableBehavior;
import akka.actor.typed.javadsl.Actor.Receive;

public class Greeter1 extends MutableBehavior<Greeter1.Command> {

  public static Behavior<Command> greeterBehavior() {
    return Actor.mutable(ctx -> new Greeter1());
  }

  interface Command {
  }

  public static class Greet implements Command {
  }

  public static class WhoToGreet implements Command {
    public final String who;

    public WhoToGreet(String who) {
      this.who = who;
    }
  }


  private String greeting = "hello";

  @Override
  public Receive<Command> createReceive() {
    return receiveBuilder()
      .onMessage(WhoToGreet.class, this::onWhoToGreet)
      .onMessage(Greet.class, this::onGreet)
      .build();
  }

  private Greeter1 onWhoToGreet(WhoToGreet whoToGreet) {
    greeting = "hello, " + whoToGreet.who;
    return this;
  }

  private Greeter1 onGreet(Greet greet) {
    System.out.println(greeting);
    return this;
  }

}